<H1
    >Palmer Penguin Data Analytics

<h1> Name: Oliver O. Herrera </h1>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')


<p>Importing dataset to Jupyter notebook </p>
<p>Converting dataset to to dataframe </p>
<ul><strong>Requirements</strong>
    <li>name of the dataframe </li>
    <li>Source </li>
</ul>

In [2]:
penguins = pd.read_csv('datasets//penguins.csv')

In [3]:
penguins


species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h3> Descriptive Analytics </h3>

In [4]:
penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h3> Cleaning the Data Frame </h3>

In [5]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h3> Syntax </h3>

<h3> data_frame['column_name']

In [6]:
penguins['island'] == 'Dream'

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: island, Length: 344, dtype: bool

In [7]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [8]:
penguins['species'].unique()

array(['Adelie', 'Gentoo', 'Chinstrap'], dtype=object)

In [9]:
penguins['island']

0      Torgersen
1      Torgersen
2      Torgersen
3      Torgersen
4      Torgersen
         ...    
339        Dream
340        Dream
341        Dream
342        Dream
343        Dream
Name: island, Length: 344, dtype: object

In [10]:
penguins['species'] == "Chinstrap"

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: species, Length: 344, dtype: bool

In [11]:
chinstrap_df = penguins[penguins['species'] == "Chinstrap"]

In [12]:
chinstrap_df

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [13]:
penguins['body_mass_g'] >= 2000

0       True
1       True
2       True
3      False
4       True
       ...  
339     True
340     True
341     True
342     True
343     True
Name: body_mass_g, Length: 344, dtype: bool

In [14]:
penguins[penguins['body_mass_g'] >= 2000]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
4       Adelie  Torgersen            36.7           19.3              193.0   
5       Adelie  Torgersen            39.3           20.6              190.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
4         3450.0  female  2007  
5         3650.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[342 rows x 8 columns]

In [15]:
(penguins['species'] == "Chinstrap") & (penguins['body_mass_g'] >= 2000)

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Length: 344, dtype: bool

In [16]:
penguins[(penguins['species'] == "Chinstrap") & (penguins['body_mass_g'] >= 2000)]

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [17]:
penguins[(penguins['species'] == "Chinstrap") & (penguins['body_mass_g'] >= 2000)].info()

<class 'pandas.core.frame.DataFrame'>
Index: 68 entries, 276 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            68 non-null     object 
 1   island             68 non-null     object 
 2   bill_length_mm     68 non-null     float64
 3   bill_depth_mm      68 non-null     float64
 4   flipper_length_mm  68 non-null     float64
 5   body_mass_g        68 non-null     float64
 6   sex                68 non-null     object 
 7   year               68 non-null     int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 4.8+ KB


In [18]:
chinstrap_heavy = penguins[(penguins['species'] == "Chinstrap") & (penguins['body_mass_g'] >= 2000)]

In [19]:
chinstrap_heavy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 68 entries, 276 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            68 non-null     object 
 1   island             68 non-null     object 
 2   bill_length_mm     68 non-null     float64
 3   bill_depth_mm      68 non-null     float64
 4   flipper_length_mm  68 non-null     float64
 5   body_mass_g        68 non-null     float64
 6   sex                68 non-null     object 
 7   year               68 non-null     int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 4.8+ KB


<h1> HOW TO FILL UP NULL VALUES </h1>

In [20]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [21]:
penguins['bill_length_mm']

0      39.1
1      39.5
2      40.3
3       NaN
4      36.7
       ... 
339    55.8
340    43.5
341    49.6
342    50.8
343    50.2
Name: bill_length_mm, Length: 344, dtype: float64

In [22]:
penguins['bill_length_mm'].mean()

np.float64(43.9219298245614)

In [23]:
penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

0      39.10000
1      39.50000
2      40.30000
3      43.92193
4      36.70000
         ...   
339    55.80000
340    43.50000
341    49.60000
342    50.80000
343    50.20000
Name: bill_length_mm, Length: 344, dtype: float64

In [24]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

In [25]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [26]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

In [27]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
dtype: int64

In [28]:
penguins['sex'].mode()

0    male
Name: sex, dtype: object

<h1> How to Change the Data Type of a Column </h1>

In [29]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     344 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                344 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


In [30]:
penguins['species'] = penguins['species'].astype('category')
penguins['year'] = penguins['year'].astype('category')

In [31]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    category
dtypes: category(2), float64(4), object(2)
memory usage: 17.2+ KB


In [32]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      344.000000     342.000000         342.000000   342.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.443643       1.974793          14.061714   801.954536
min         32.100000      13.100000         172.000000  2700.000000
25%         39.275000      15.600000         190.000000  3550.000000
50%         44.250000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000

<h1> How to create a column </h1>

In [33]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    category
dtypes: category(2), float64(4), object(2)
memory usage: 17.2+ KB


In [34]:
penguins.columns

Index(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'],
      dtype='object')

In [35]:
penguins['mass_kg'] = np.round(penguins['body_mass_g'] / 5000,1)

In [36]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000           18.7              181.0   
1       Adelie  Torgersen        39.50000           17.4              186.0   
2       Adelie  Torgersen        40.30000           18.0              195.0   
3       Adelie  Torgersen        43.92193            NaN                NaN   
4       Adelie  Torgersen        36.70000           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000           19.8              207.0   
340  Chinstrap      Dream        43.50000           18.1              202.0   
341  Chinstrap      Dream        49.60000           18.2              193.0   
342  Chinstrap      Dream        50.80000           19.0              210.0   
343  Chinstrap      Dream        50.20000           18.7              198.0   

     body_mass_g     sex  year  mass_kg  
0         3750.0    male  2007      0.8  
1         3800.0  female  2007      0.8  
2         3250.0  female  2007      0.6  
3            NaN    male  2007      NaN  
4         3450.0  female  2007      0.7  
..           ...     ...   ...      ...  
339       4000.0    male  2009      0.8  
340       3400.0  female  2009      0.7  
341       3775.0    male  2009      0.8  
342       4100.0    male  2009      0.8  
343       3775.0  female  2009      0.8  

[344 rows x 9 columns]

<h1> Exploratory Data Analytics (EDA) </h1>

<h3> 1. Which species has the heaviest body mass in kg? </h3>

In [37]:
chinstrap_df['body_mass_g'].mean()

np.float64(3733.0882352941176)

In [38]:
adelie_df = penguins[penguins['species'] == 'Adelie']

In [39]:
adelie_df['body_mass_g'].mean()

np.float64(3700.662251655629)

In [40]:
gentoo_df = penguins[penguins['species'] == 'Gentoo']

In [41]:
gentoo_df['body_mass_g'].mean()

np.float64(5076.016260162602)

<h2> Insight 1 - 7 Questions </h2>

<h2>1. Which island has the highest number of penguins?

In [55]:
penguins['island'].value_counts()

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

<h3> Insight: Biscoe Island has the most penguins recorded (168), followed by Dream (124) and then Torgersen(52).

<h2> 2. What is the number of male vs female penguins?

In [56]:
penguins['sex'].value_counts()

sex
male      179
female    165
Name: count, dtype: int64

<h3> Insight: Male penguins are slightly more than female penguins 

<h2>3. What is the average flipper length for each species?

In [57]:
penguins.groupby("species")["flipper_length_mm"].mean()

species
Adelie       189.953642
Chinstrap    195.823529
Gentoo       217.186992
Name: flipper_length_mm, dtype: float64

<h3> Insight: Gentoo penguins have longer flippers (~217mm)

<h2>4. What is the average bill length by sex?

In [60]:
penguins.groupby("sex")["bill_length_mm"].mean()

sex
female    42.096970
male      45.604156
Name: bill_length_mm, dtype: float64

<h3> Insight: Male penguins generally have longer bills than females.

<h2>5. What is the average body mass per species?

In [62]:
penguins.groupby("species")["body_mass_g"].mean()

species
Adelie       3700.662252
Chinstrap    3733.088235
Gentoo       5076.016260
Name: body_mass_g, dtype: float64

<h3> Insight: Gentoo penguins are significantly heavier on average than Adelie and Chinstrap penguins.

<h2>6. Which species has the widest average bill?

In [63]:
penguins.groupby("species")["bill_depth_mm"].mean()

species
Adelie       18.346358
Chinstrap    18.420588
Gentoo       14.982114
Name: bill_depth_mm, dtype: float64

<h3> Insight: Chinstrap penguins have the widest bills on average, followed closely by Adelie. Gentoo penguins have noticeably narrower bills

 <h2>7. What is the distribution of species by island?

In [64]:
pd.crosstab(penguins['island'], penguins['species'])

species    Adelie  Chinstrap  Gentoo
island                              
Biscoe         44          0     124
Dream          56         68       0
Torgersen      52          0       0

<h3> Insight: Gentoo penguins are only found on Biscoe Island, while Chinstrap penguins are only on Dream Island but the Adelie penguins are distributed across all three islands

<h2>8. How many missing values are there in each column?

In [66]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
mass_kg              2
dtype: int64

<h3> Insight: Most missing data is in the sex column (11 entries)

<h2> 9. What is the correlation between flipper length and body mass?

In [69]:
penguins[['flipper_length_mm', 'body_mass_g']].corr()

flipper_length_mm  body_mass_g
flipper_length_mm           1.000000     0.871202
body_mass_g                 0.871202     1.000000

<h3> Insight: There is a strong positive correlation (0.87) between flipper length and body mass.

<h2> 10. What is the average flipper length per sex?



In [71]:
penguins.groupby('sex')['flipper_length_mm'].mean()


sex
female    197.363636
male      204.225989
Name: flipper_length_mm, dtype: float64

<h3> Insight: Male penguins have longer flippers on average than females.

<h2> 11. Male penguins have longer flippers on average than females.

In [73]:
penguins.groupby("sex")["bill_length_mm"].mean()


sex
female    42.096970
male      45.604156
Name: bill_length_mm, dtype: float64

<h3> Insight: Males have longer bills on average than females.

<h2> 12. How many penguins belong to each species?

In [74]:
penguins["species"].value_counts()


species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

<h3> Insight: Adelie is the most represented species in the dataset, followed by Gentoo, then Chinstrap.